### DISCUSSION TITLE
Tesvor x500

### DISCUSSION THREAD

**Discussion by Sharkozp** [arendst/Tasmota#19554] [Show and tell]
Hello there,

I found a similar device to my Tesvor x500 in your collection:
[Dealdig Robvaccum 8 Vacuum](https://templates.blakadder.com/dealdig_Robvaccum_8.html)

I decided to try installing Tasmota as well. After opening up the device, I found an ESP-12S module. I successfully installed Tasmota and used the template from there:
`{"NAME":"WhiteVacuum","GPIO":[0,107,0,108,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":18}`

The only difference I found is that the device sends different serial commands
My is:
`AA1C060000000500020000000000000000000000000000000000000000424F`
Dealdig Robvaccum 8 Vacuum:
`AA1C06FF00040500000000000000000000000000000000000000000000505E`

Right now I can controll Tesvor X500 from Home Assistant.

Also my device is sending this, when it is working:
`{"SerialReceived":"AA0E0301400000010B690040004003003CAA02030508"}`
It seems somewhat similar to the map-building feature in the WeBack app

**Comment by Gatono27** [comment]
Buen dia, por casualidad quemé mi esp12s, alguna forma de reemplazarlo?
Que me recomiendas hacer?

**Comment by DasRed** [comment]
Can you give us a more detail description, how exactly you did that?

**Reply by calipee** [reply]
I didn't manage to remove the build in ESP-12S and just added a second one flashed with Tasmota by connecting the cables. 
What do you want to know exactly?

**Comment by lovicub** [comment]
@DasRed 

Hey,

I've also managed to get a TESVOR X500 up and running using this method. Here's what I did:

1. I got myself a Node esp12 Flasher with pogopins so I wouldn't have to solder, and I flashed Tasmota generic onto it via browser. 

![image](https://github.com/arendst/Tasmota/assets/29915644/b6a114ae-3050-41cb-b06d-1e359a084734)

(it was a pain, bacause I had to bend the pins slightly to ensure proper contact - I tried for about an hour, but I succeeded somehow.)

![image](https://github.com/arendst/Tasmota/assets/29915644/89408238-a604-49ad-95b7-1f9b087c88d1)


3. Then, I connected the ESP module to my WLAN.

4.Next, I reinstalled the module into the X500 and turned on the vacuum cleaner. After that, I went into the configuration and included the YAML file provided here. I conducted a few tests in the Tasmota console:

start / stop
SerialSend5 0,HEX,aa0302220226
SerialSend5 0,HEX,aa0302260028

![image](https://github.com/arendst/Tasmota/assets/29915644/625fb9a3-b34b-4024-b99b-aa3495c97ddb)
![image](https://github.com/arendst/Tasmota/assets/29915644/803d2825-1dcb-442a-81d2-0b42ebe806da)

I noticed that I could control the X500 via the "Vacuum Custom Card" but couldn't display status queries in HA. Upon further investigation in the console, I found that other SerialReceived (tele/WhiteVacuum/RESULT) codes were being output. I analyzed these and found the following differing values, for example:

AA1C0600000005 idle
AA1C0600000002 cleaning

Despite adjusting the YAML, there still wasn't a proper display of what the vacuum cleaner was currently doing.

Since I didn't have much experience with YAML and virtual sensors, I brought them into Node-RED and simply provided the sensors here with the correct statuses.
![image](https://github.com/arendst/Tasmota/assets/29915644/1d94be0b-4566-4f38-b1f8-9d9d96744de1)

After that, it worked flawlessly with all statuses. (Except for minimal and maximal fan speed).

![image](https://github.com/arendst/Tasmota/assets/29915644/a7e89ac7-a261-430b-b064-cf1541747a51)

here function for battery calculation in nodered function;

```
var serialReceived = msg.payload.SerialReceived;
msg.payload.serialReceived = serialReceived;

var msg = {
    payload: {} // Initializing msg.payload
};

// Extract the last 4 characters of the serialReceived string
var lastFourCharacters = serialReceived.slice(-4);
msg.payload.lastFourCharacters = lastFourCharacters;
// Extract the first 2 characters of the last four characters
var batteryHex = lastFourCharacters.slice(0, 2);
msg.payload.batteryHex = batteryHex;

// Convert the hexadecimal string to a decimal number
var batteryDecimal = parseInt(batteryHex, 16);
msg.payload.batteryDecimal = batteryDecimal;

// Now batteryDecimal is the battery value in decimal format
var batteryString = batteryDecimal.toString();
msg.payload.batteryString = batteryString;

return msg;
```


use batteryDecimal for the sensor.whitevacuum_battery

**Reply by JRDevo** [reply]
I have a Tesvor x500 rebranded to 'cleanwise' which was imported into the Netherlands. The company no longer exists.

@lovicub would it be possible to expand the steps in more detail please? Specifically "Then, I connected the ESP module to my WLAN."? with some pics.

Thanks in advance :)

**Reply by sfromis** [reply]
When Tasmota has [been installed](https://tasmota.github.io/docs/Getting-Started/), that step is simply the standard step of connecting Tasmota to Wifi, typically by connecting to the initial AP, and selecting your network.

What the screenshots above does not make clear is what the template is, only a small part of it is shown, but since gpio 1 is `Tuya Tx`, gpio 3 is most probably `Tuya Rx`, and based on module `Tuya MCU (54)`. Actually, if you just select the based on module, it will default to gpio 1/3 for Tx/Rx.

**Reply by devasworski** [reply]
Hey, thank you so much for posting this.
I had been looking for something like this for a while. I connected the three pins to a logic analyser a while back, but couldn't yet figure out some of the signals.
I have two quick questions:

1. How did you remove the original WiFi module? Mine is glued and I can't move it at all.
2. Did you flash the original WiFi module or did you attach a new ESP 12S? Aka, do you see any issues doing it with a ESP32 on a breadboard instead? 

My ultimate goal is to replace the WiFi module with a ESP32 with Thread Support to make it Matter enabled. But installing Tasmota first, analysing that a bit more and then going the Matter route seems like a good idea right now. I will link my progress once I have something worth sharing.

**Reply by sfromis** [reply]
You can also get ESP32-based modules in the same form factor as ESP-12, reusing the same small green board. But if that 4-pin connector is all the connections, you can do whatever, as long as it fits. Beware that ESP32 needs more power, and it is not a given that this will be available.

**Reply by devasworski** [reply]
So you reused the original board. Do you know if that is necessary? Is there any additional logic or components on the PCB that are required for this to work?  As I said, I didn't manage to get mine out of the case as it is heavily glued.

**Reply by sfromis** [reply]
It was not me having the hardware, I just posted a general bit of info about potential strategies for replacement. When it comes to the specifics, I already mentioned that if all connections are the 4-pin connector, that's what you can act on if you connect the cable from the main board to something else instead.

As the connections from the ESP-12 module to the 4-pin connector are partly hidden, I can't give any details of exactly how the board works, how the signals to the main board acts, but since there are no active components, I'd not expect too much trouble.

**Reply by devasworski** [reply]
Did you experience any issues with the robot not going back to its base, but just continuing to vacuum?

**Comment by dammert89** [comment]
This helped me so much! Thanks @lovicub I did the exact same with an esp32 devkit. 

So far I have figured out the following codes:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/MICHAE~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/MICHAE~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


ACTION | SERIAL RECEIVED | SERIAL SEND
-- | -- | --
MOVEFORWARD | AA1C06FF00000A | AA0302210023AA0302210427
TURNLEFT | AA1C06FF00010A | AA0302210124AA0302210427
TURNRIGHT | AA1C06FF00020A | AA0302210225AA0302210427
MOVEBACK | AA1C06FF00030A | AA0302210326AA0302210427
BORDERCLEANINGMODE | AA1C06FF000400 | AA0302220024
FIXEDPOINTCLEANINGMODE | AA1C06FF000401 | AA0302220125
SMARTCLEANINGMODE | AA1C06FF000402 | AA0302220226
GOCHARGE | AA1C06FF000403 | AA0302220327
IDLE/STOP | AA1C06FF000404 | AA0302260028
CHARGINGSTART | AA1C06FF000406 |  



</body>

</html>

These are based on my Dealdig Robot Vaccuum 8.
I am still trying to find the fast/slow fan mode however I have had no luck on that.

**Comment by thomasklesse-cloud** [comment]
@dammert89 and @lovicub 
Thank you for sharing your knowledge.
I recently noticed that my robot had lost its network connection and it was time to disconnect it from all Chinese clouds. 
Yes, I know this project is very old, but today I wanted to let everyone know how to reuse the good robot in Home Assistant in the future. 

The links and sources are a bit confusing. The YAML file contains everything, the vacuum cleaner configuration, but also scripts and automation tags. These examples are useful, but incorrect in this one file.

I flashed an ESP-8266 board with Tasmota and replaced it with the original. The pin assignment of the connector is from left to right: 3.3 V RX/TX TX/RX GND. RX / TX must be crossed!The scripts work wonderfully, but for my Tesvor X500, the only moving configuration does not work. The solution is instead of, for example:

TURNLEFT: Send AA0302210124AA0302210427

, use only
TURNLEFT: Send AA0302210124 in the Script/Yaml-File.

**Reply by lovicub** [reply]
this worked for my configuration as a HA script
front:
```
data_template:
  topic: cmnd/WhiteVacuum/SerialSend5
  payload: aa 03 02 21 00 23 aa 03 02 21 04 27
action: mqtt.publish
```
back:
`  payload: aa 03 02 21 03 26 aa 03 02 21 04 27`
left:
`  payload: aa 03 02 21 02 25 aa 03 02 21 04 27`
right:
`  payload: aa 03 02 21 01 24 aa 03 02 21 04 27`

**Comment by darwinwieting4413-sudo** [comment]
I found a solution using ESPHome. It works better than Tasmota because Tasmota doesn't understand hex code and therefore doesn't show any output in the log, whereas ESPHome does. I swapped the ESP12S for an ESP8266 D1 Mini V4. Take a look here: https://github.com/hannes813/tesvor-x500-pro-esphome-vacuum/blob/main/x500.yaml  hannes813 has published a great ESPHome configuration for Home Assistant that even includes map visualization.

Wired as follows:

GND to GND

tx_pin: GPIO1                                  

rx_pin: GPIO14

3V3 to 3V3 (not to Vbus)
